src/SymfonyCache/KernelDispatcher.php

```diff
@@ -0,0 +1,115 @@
+<?php
+
+/*
+ * This file is part of the FOSHttpCache package.
+ *
+ * (c) FriendsOfSymfony <http://friendsofsymfony.github.com/>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace FOS\HttpCache\SymfonyCache;
+
+use FOS\HttpCache\Exception\ExceptionCollection;
+use FOS\HttpCache\ProxyClient\Dispatcher;
+use Psr\Http\Message\RequestInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\HttpKernelInterface;
+
+/**
+ * An implementation of Dispatcher that allows direct calling of the
+ * Symfony HttpCache kernel without executing a real HTTP request.
+ * It can only be used if you have a single node setup of Symfony and serves
+ * as kind of a shortcut for easier configuration.
+ * If you use Varnish or have a multiple node Symfony setup, this client is entirely
+ * useless and cannot be used.
+ *
+ * @author Yanick Witschi <yanick.witschi@terminal42.ch>
+ */
+class KernelDispatcher implements Dispatcher
+{
+    /**
+     * @var HttpCacheProvider
+     */
+    private $kernel;
+
+    /**
+     * @var array
+     */
+    private $queue = [];
+
+    /**
+     * KernelClient constructor.
+     *
+     * @param HttpCacheProvider $kernel
+     */
+    public function __construct(HttpCacheProvider $kernel)
+    {
+        $this->kernel = $kernel;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function invalidate(RequestInterface $invalidationRequest, $validateHost = true)
+    {
+        $request = Request::create(
+            $invalidationRequest->getUri(),
+            $invalidationRequest->getMethod(),
+            [],
+            [],
+            [],
+            ['REMOTE_ADDR' => '127.0.0.1'],
+            $invalidationRequest->getBody()->getContents()
+        );
+
+        // Add headers
+        $headers = $invalidationRequest->getHeaders();
+        foreach ($headers as $name => $values) {
+            $name = strtolower($name);
+
+            if ('cookie' === $name) {
+                foreach ($values as $value) {
+                    foreach (explode(';', $value) as $cookieString) {
+                        $chunks = explode('=', $cookieString, 2);
+                        $request->cookies->add([trim($chunks[0]) => trim($chunks[1])]);
```

Comment:
if we simply set the cookie header like the other headers, is symfony overwriting it?

Reply:
The `Request` class of Symfony unfortunately has no handling for `cookie` in the request class. It expects you to set them from the outside (usually `$_COOKIE`) so I had to implement custom handling here. I kind of expected that to happen which is why I put it into the unit test right from the beginning :D

Reply:
on http level, cookies is just another header. is setting that header destroying things? (i guess we do care, because `refresh` in combination with user context - although that would be a bit random, it would only refresh for the group of the client triggering the refresh)

Reply:
i would have expected that setting the cookie header would work, but i see, this is parsed from the $_COOKIE only. i think now we do the correct thing, convert the $invalidationRequest cookies into symfony request cookies so they are processed :+1: